paasta_tools/setup_kubernetes_job.py

```diff
@@ -140,11 +144,21 @@ def setup_kube_deployments(
             not in existing_apps
         ):
             app.create(kube_client)
-        elif app and app.kube_deployment not in existing_kube_deployments:
+        elif (
+            app
+            and (app.kube_deployment not in existing_kube_deployments)
+            or autoscaling_is_paused()
+        ):
+            app.update(kube_client)
+        elif not autoscaling_is_paused() and not is_autoscaling_resumed():
+            resumed_autoscaler = True
             app.update(kube_client)
         else:
             log.debug(f"{app} is up to date, no action taken")
 
+    if resumed_autoscaler:
+        write_autoscaling_resumed()
+
```

Comment:
This elif is the reason for the existence of `is_autoscaling_resumed`, right? We're trying to avoid updating all k8s deployments whenever the autoscaler is paused?

I worry about the resumed key and the paused key getting out of sync somehow. Is there a way to tell the last time the k8s deployment was updated, and compare that to the pause expiration timestamp?

I also worry about `setup_kubernetes_job` getting called with some subset of services. (At the moment, we call SKJ with all service instances with `paasta_list_kubernetes_service_instances | shuf | xargs setup_kubernetes_job`, but for Marathon before we had deployd we parallelized `setup_marathon_job` with `xargs --max-procs`. I could imagine us doing this if SKJ starts taking too long to run.)

If called with a subset of services, we'd still update the resumed key in ZK and then any services that we excluded would never get their HPA reset.

I think comparing to the pause expiration timestamp would avoid this issue too.

Reply:
@EvanKrall Hmm good point. One way we could fix this is adding an annotation to all deployments that have been paused. I think this might be better than checking the last time the k8s deployment was updated, because deployments could possibly be updated for multiple reasons outside of being paused.

Also by adding annotations to each in individual deployment, we could still support calling `setup_kubernetes_job` for a subset of services.